#516375 [C#テキスト] 1章演習問題　

sanada-being opened the pull request:
演習問題の回答を作成しました。

sanada-being committed:
#516375 [C#テキスト] 1章演習問題

gitignoreとChapter1のプロジェクトを追加しました。

sanada-being committed:
#516375　[C#テキスト] 1章演習問題
VSの設定

sanada-being committed:
#516375  [C#テキスト] 1章演習問題

演習問題の回答を作成しました。

teshima-being commented on Chapter1/Chapter1-1-1/Program.cs:
```diff
@@ -0,0 +1,36 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
不要なusingは削除しておきましょう。
Visual Studio で確認した際にグレー表示になっているものは使用されていません。

teshima-being commented on Chapter1/Chapter1-1-1/Program.cs:
```diff
@@ -0,0 +1,36 @@
+        //　この時の商品番号は"98"、商品価格は"210円"としてください。
+        //2.どら焼きオブジェクトの消費税額を求め、コンソールに出力するコードを書いてください。
+        //3.Productクラスが属する名前空間を別の名前空間に変更し、Mainメソッドから呼び出すようにしてください。
+        //ただし、MainメソッドのあるProgramクラスの名前はそのままとしてください。
```
`//`とコメント本文の間には半角スペースを1つ挟むようにしましょう。可読性向上のためです。

※以下修正不要
コメントには`//`を使用するシングルラインコメントと、`/*`で開始し、`*/`で終了することで、複数行に渡ってコメントを記述できるマルチラインコメントがあります。どちらも知っておくとよいでしょう。

teshima-being commented on Chapter1/Chapter1-1-1/Product.cs:
```diff
@@ -0,0 +1,31 @@
+        //「1-1クラス」で定義したProductクラス
+        public int Code { get; private set; }
+        public string Name { get; private set; }
+        public int Price { get; private set; }
```
クラスやプロパティ、コンストラクタ、メソッドには[ドキュメントコメント](https://learn.microsoft.com/ja-jp/visualstudio/ide/reference/generate-xml-documentation-comments?view=vs-2022)をつけましょう。
そうすることにより、そのクラスやプロパティなどにマウスをホバーした際に、記述したコメントが確認できるようになります。
![スクリーンショット 2024-10-15 110650](https://github.com/user-attachments/assets/e1af69ff-ae0f-44de-9112-9c83c2e80bfa)
クラス外からプロパティなどを使うときに、そのプロパティが定義されているクラスにわざわざ移動しなくても、どういったプロパティなのかをひと目で確認できるようになるため、非常に便利です。

ちなみに、クラスやプロパティなどのすぐ上の行で`///`と入力すると、自動で補完してくれます。

teshima-being commented on Chapter1/Chapter1-1-1/Program.cs:
```diff
@@ -0,0 +1,36 @@
+            int wDorayakiTax = wDorayaki.GetTax();
+            Console.WriteLine(wDorayakiTax);
+
+
```
細かいですが、余分な改行は削除しておきましょう。指摘の箇所だと1行あれば十分かと思います。
他にもいくつか余分な改行がありますので、一度全体をチェックしてみてください。

teshima-being commented on Chapter1/Chapter1-1-1/Program.cs:
```diff
@@ -0,0 +1,36 @@
+
+            //2.どら焼きオブジェクトの消費税額を求め、コンソールに出力するコードを書いてください。
+            int wDorayakiTax = wDorayaki.GetTax();
+            Console.WriteLine(wDorayakiTax);
```
現在は消費税額に相当する金額のみがコンソールに表示されるようになっていますが、数字のみだと何を表しているのか分かりにくいので、"消費税額は〇〇円です。" のようにコンソールに表示するようにしてみてください。

teshima-being commented on Chapter1/Chapter1-1-1/Program.cs:
```diff
@@ -0,0 +1,36 @@
+
+            //3.変更した名前空間のメソッドをmainメソッドから呼び出し
+            int wTaxIncluded = wDorayaki.GetPriceIncludingTax();
+            Console.WriteLine(wTaxIncluded);
```
ローカル変数の`wDorayakiTax`や`wTaxIncluded`などは一度しか使用されておらず、特に宣言する必要性がありません。
このような場合は、`Console.WriteLine`の引数に直接メソッドの戻り値を指定してあげましょう。
例) `Console.WriteLine（wDorayaki.GetPriceIncludingTax());`

teshima-being commented on Chapter1/Chapter1-1-3/Program.cs:
```diff
@@ -0,0 +1,45 @@
+            //4.Person型、object型の変数にそれぞれ代入
+            Person wPerson = wStudent1;
+            object wObject = wStudent1;
+            Console.WriteLine( wPerson.Name);
```
細かいですが、余分な半角スペースがありますので削除しておきましょう。
ちなみに、Visual Studio で「編集」> 「詳細」>「スペースの表示」をすることで、スペースが見える化されます。

また、「Ctrl + K」>「Ctrl + D」の順に押下することで、余計な空白を削除してくれたり、インデントを整えてくれたりします。（「Ctrl + K」>「Ctrl + F」で選択範囲のコード整形も可能です。）

teshima-being commented on Chapter1/Chapter1-1-3/Person.cs:
```diff
@@ -0,0 +1,20 @@
+
+namespace Chapter1_1_3 {
+    internal class Person {
+        public String Name { get; set; }
```
String表記とstring表記については機能的に差がないようですが、Microsoftではstringの仕様が推奨されているようです。stringにしておきましょう。
参考：[文字列と System.String](https://learn.microsoft.com/ja-jp/dotnet/csharp/programming-guide/strings/#:~:text=C%23%20%E3%81%A7%E3%81%AF%E3%80%81string%20%E3%82%AD%E3%83%BC%E3%83%AF%E3%83%BC%E3%83%89%E3%81%AF%20String%20%E3%81%AE%E3%82%A8%E3%82%A4%E3%83%AA%E3%82%A2%E3%82%B9%E3%81%A7%E3%81%99%E3%80%82%20%E3%81%97%E3%81%9F%E3%81%8C%E3%81%A3%E3%81%A6%E3%80%81String%20%E3%81%A8%20string%20%E3%81%AF%E5%90%8C%E7%AD%89%E3%81%A7%E3%81%99%E3%80%82%E3%81%9F%E3%81%A0%E3%81%97%E3%80%81using%20System%3B%20%E3%81%8C%E3%81%AA%E3%81%8F%E3%81%A6%E3%82%82%E6%A9%9F%E8%83%BD%E3%81%99%E3%82%8B%E3%81%93%E3%81%A8%E3%81%8B%E3%82%89%E3%80%81%E6%8C%87%E5%AE%9A%E3%81%95%E3%82%8C%E3%81%9F%E3%82%A8%E3%82%A4%E3%83%AA%E3%82%A2%E3%82%B9%20string%20%E3%82%92%E4%BD%BF%E7%94%A8%E3%81%99%E3%82%8B%E3%81%93%E3%81%A8%E3%82%92%E6%8E%A8%E5%A5%A8%E3%81%97%E3%81%BE%E3%81%99%E3%80%82%20String%20%E3%82%AF%E3%83%A9%E3%82%B9%E3%81%AF%E3%80%81%E6%96%87%E5%AD%97%E5%88%97%E3%82%92%E5%AE%89%E5%85%A8%E3%81%AB%E4%BD%9C%E6%88%90%E3%80%81%E6%93%8D%E4%BD%9C%E3%80%81%E6%AF%94%E8%BC%83%E3%81%99%E3%82%8B%E3%81%9F%E3%82%81%E3%81%AE%E3%81%95%E3%81%BE%E3%81%96%E3%81%BE%E3%81%AA%E3%83%A1%E3%82%BD%E3%83%83%E3%83%89%E3%82%92%E6%8F%90%E4%BE%9B%E3%81%97%E3%81%BE%E3%81%99%E3%80%82%20%E3%81%BE%E3%81%9F%E3%80%81C%23%20%E8%A8%80%E8%AA%9E%E3%81%AF%E3%80%81%E4%B8%80%E8%88%AC%E7%9A%84%E3%81%AA%E6%96%87%E5%AD%97%E5%88%97%E6%93%8D%E4%BD%9C%E3%82%92%E7%B0%A1%E7%95%A5%E5%8C%96%E3%81%99%E3%82%8B%E3%81%9F%E3%82%81%E3%81%AB%E3%81%84%E3%81%8F%E3%81%A4%E3%81%8B%E3%81%AE%E6%BC%94%E7%AE%97%E5%AD%90%E3%82%92%E3%82%AA%E3%83%BC%E3%83%90%E3%83%BC%E3%83%AD%E3%83%BC%E3%83%89%E3%81%97%E3%81%BE%E3%81%99%E3%80%82%20%E3%82%AD%E3%83%BC%E3%83%AF%E3%83%BC%E3%83%89%E3%81%AE%E8%A9%B3%E7%B4%B0%E3%81%AB%E3%81%A4%E3%81%84%E3%81%A6%E3%81%AF%E3%80%81%E3%80%8Cstring%E3%80%8D%E3%82%92%E5%8F%82%E7%85%A7%E3%81%97%E3%81%A6%E3%81%8F%E3%81%A0%E3%81%95%E3%81%84%E3%80%82%20%E5%9E%8B%E3%81%8A%E3%82%88%E3%81%B3%E3%83%A1%E3%82%BD%E3%83%83%E3%83%89%E3%81%AE%E8%A9%B3%E7%B4%B0%E3%81%AB%E3%81%A4%E3%81%84%E3%81%A6%E3%81%AF%E3%80%81%E3%80%8CString%E3%80%8D%E3%82%92%E5%8F%82%E7%85%A7%E3%81%97%E3%81%A6%E3%81%8F%E3%81%A0%E3%81%95%E3%81%84%E3%80%82)

teshima-being commented on Chapter1/Chapter1-1-3/Person.cs:
```diff
@@ -0,0 +1,20 @@
+        public int GetAge() {
+            DateTime wToday = DateTime.Today;
+            int wAge = wToday.Year - Birthday.Year;
+            if (wToday < Birthday.AddYears(wAge))
+                wAge--;
```
※以下修正は任意
テキスト自体がこのようになっていた覚えがありますが、if 文内の処理が1行ですむ場合は、
`if (wToday < this.Birthday.AddYears(wAge)) wAge--;`
のように条件判定と処理を1行で記述することができます。
それ以外の場合については{ }を省略しない方が、分岐による処理がどこまでなのかが明確になり、のちのち意図せぬバグを埋め込むリスクが下がります。

例えば、
```
if (wToday < this.Birthday.AddYears(wAge))
　wAge--;　// ①
　Console.WriteLine("今年はまだ誕生日が来ていません");　// ②
```
のようにすると、①は if 文の条件判定が true の場合のみ実行されますが、②は if 文の条件判定に関係なく常に実行されてしまいます。

teshima-being commented on Chapter1/Chapter1-1-3/Program.cs:
```diff
@@ -0,0 +1,45 @@
+        static void Main(string[] args) {
+
+
+            //2.すべてのプロパティに値を入れたインスタンスを生成するコード
+            Student wStudent1 = new Student {
+                Name = "Akbar Shah Sanada",
+                Birthday = new DateTime(2001, 11, 13),
+                Grade = 6,
+                ClassNumber = 1,
+            };
```
せっかくなので、Studentクラスにコンストラクタを実装し、実装したコンストラクタを利用してプロパティに値を設定してみてください。

teshima-being commented on Chapter1/Chapter1-1-3/Program.cs:
```diff
@@ -0,0 +1,45 @@
+            };
+            int wAge = wStudent1.GetAge();
+
+            // 3.で生成したインスタンスの各プロパティの値をコンソールに出力するコード
+            Console.WriteLine("{0}は{1}年{2}組に所属しており、生年月日は{3}です。{4}歳です",
+                wStudent1.Name, wStudent1.Grade, wStudent1.ClassNumber, wStudent1.Birthday, wAge);
```
コンソールへの出力の際には、文字列補間式を使ってみてください。
参考：[C# 文字列の文字列補間](https://csharp.keicode.com/basic/string-interpolation.php)
　　　[[C#] 文字列の書式設定を行う方法](https://zenn.dev/rinc5/articles/8e1e68f56993ce)

他の箇所でも同様に文字列補間式で出力してみましょう。

teshima-being replied:
現在は生年月日について、"生年月日は 2001/11/13 00:00:00です" と出力されますが、余分な時間は表示させずに"生年月日は2001年11月13日です" のように表示させてみてください。

teshima-being reviewed (commented):
レビューしましたので、ご確認ください。
指摘された事項については、他の箇所でも同様に修正すべき箇所がないか確認して、必要に応じて修正するようお願いします。

sanada-being committed:
タスク #516375   [C#テキスト] 1章演習問題

・不要なusingを削除しました。
・//とコメント本文に半角スペースを挿入しました。
・ドキュメントコメントを追加しました。
・余分な改行を削除しました。
・余分な半角スペースを削除しました。
・コンソールへの出力の際に文字列補間式で出力するよう変更しました。
・Chapter1/Chapter1-1-1/Program.cs22行目にて、消費税額のみの表示から、"消費税額は○○です。"に変更しました。
Chapter1/Chapter1-1-1/Program.cs22,25行目にて、Console.WriteLineの引数に直接メソッドの戻り値を指定するように変更しました。
・Chapter1/Chapter1-1-3/Person.cs11行目にて、Stringをstring表記に修正しました。
・Chapter1/Chapter1-1-3/Person.cs25行目にて、 if文を1行に修正しました。
・Chapter1/Chapter1-1-3/Program.cs24行目にて、Studentクラスにコンストラクタを作成し、コンストラクタを利用してプロパティに値を設定しました。
Chapter1/Chapter1-1-3/Program.cs24行目にて、誕生日が日付までしか表示されないように修正しました。

teshima-being commented on Chapter1/Chapter1-1-1/Product.cs:
```diff
@@ -1,29 +1,50 @@
+        /// </summary>
+        /// <returns>税額を整数値で返す。</returns>
         public int GetTax() {
             return (int)(Price * 0.08);
```
`Price`のように、クラス内のプロパティを参照するときはthisをつけるようにしましょう。
thisについては[コーディング規約](https://dev-redmine.being.group/projects/sirius-client/wiki/%E3%82%B3%E3%83%BC%E3%83%87%E3%82%A3%E3%83%B3%E3%82%B0%E8%A6%8F%E7%B4%84#:~:text=this%0A%E3%80%80%E3%80%80%E3%83%97%E3%83%AD%E3%83%91%E3%83%86%E3%82%A3%EF%BC%9A%E3%82%AF%E3%83%A9%E3%82%B9%E5%86%85%E3%81%A7%E3%83%97%E3%83%AD%E3%83%91%E3%83%86%E3%82%A3%E3%82%92%E5%8F%82%E7%85%A7%E3%81%99%E3%82%8B%E5%A0%B4%E5%90%88%E3%81%AF%20this.Property%20%E3%81%A8%E6%9B%B8%E3%81%8F%E3%81%AE%E3%82%92%E6%8E%A8%E5%A5%A8%E3%80%82%0A%E3%80%80%E3%80%80%E3%83%95%E3%82%A3%E3%83%BC%E3%83%AB%E3%83%89%EF%BC%9A%E6%8E%A5%E9%A0%AD%E8%BE%9E%E3%81%AE%20%22F%22%20%E3%81%A7%E8%87%AA%E8%BA%AB%E3%81%AE%E3%83%95%E3%82%A3%E3%83%BC%E3%83%AB%E3%83%89%E3%81%A0%E3%81%A8%E5%88%86%E3%81%8B%E3%82%8B%E3%81%AE%E3%81%A7%20this%20%E3%82%92%E4%BB%98%E3%81%91%E3%81%AA%E3%81%8F%E3%81%A6%E3%82%82%E3%81%84%E3%81%84%E3%80%82%E4%BB%98%E3%81%91%E3%81%A6%E3%82%82%E3%81%84%E3%81%84%E3%80%82%0A%E3%80%80%E3%80%80%E3%83%A1%E3%82%BD%E3%83%83%E3%83%89%EF%BC%9A%E3%80%80this%20%E3%82%92%E4%BB%98%E3%81%91%E3%81%A6%E3%82%82%E4%BB%98%E3%81%91%E3%81%AA%E3%81%8F%E3%81%A6%E3%82%82%E3%81%84%E3%81%84%E3%80%82)にも記載がありますので、一度確認してみてください。

また、Chapter1-1-3の`Person.cs`や`Student.cs`でも同様の箇所があるので、確認のほどよろしくお願いします。

teshima-being commented on Chapter1/Chapter1-1-1/Product.cs:
```diff
@@ -1,29 +1,50 @@
         public int Price { get; private set; }
 
-        //コンストラクタ
+        ///  <summary>
+        /// 商品情報初期化のためのコンストラクタ
+        /// </summary>
+        /// <param name="vCode">商品コードを示す整数値。</param>
+        /// <param name="vName">商品名を表す文字列。</param>
+        /// <param name="vPrice">商品名の価格を表す整数値。</param>
```
それぞれの引数が何を指すのか分かりやすくなりましたね！
あとは終礼時にも少し話しましたが、コードからすぐ分かること（今回の場合だと「～を示す整数値」の部分など）は冗長になるので記載しなくて良いです。
逆に今回の問題だと確か`vPrice`は「税抜」価格だったかと思いますが、価格が税抜か税込かなどが明確でない場合は、そこについての記載はある方が優しいと思います。
```
        /// <summary>
        /// コンストラクタ
        /// </summary>
        /// <param name="vCode">商品コード</param>
        /// <param name="vName">商品名</param>
        /// <param name="vPrice">税抜価格</param>
```

細かいですが、`<summary>`の前の半角スペースが1つ多いので、それも削除しておきましょう。

teshima-being commented on Chapter1/Chapter1-1-1/Program.cs:
```diff
@@ -1,36 +1,28 @@
         static void Main(string[] args) {
-            //1.どら焼きオブジェクトを生成するコードをかいてください。
+            // 1.どら焼きオブジェクトを生成するコードをかいてください。
             Product wDorayaki = new Product(98, "どら焼き", 210);
```
※参考（修正は任意）
テキストでまだ出てきていないかもしれませんが、右辺から変数の型が明らかな場合は`var`（暗黙的な型指定）を利用して、下記のように記述するとよいです。また時間があれば調べてみてください。
`var wDorayaki = new Product(98, "どら焼き", 210);`

teshima-being commented on Chapter1/Chapter1-1-1/Program.cs:
```diff
@@ -1,36 +1,28 @@
 using Chapter1_1_1change;
 
 namespace Chapter1_1_1 {
+    
```
6行目に余分な半角スペースがありますので、削除しておきましょう。（改行は問題ないです。）

teshima-being commented on Chapter1/Chapter1-1-2/MyStruct.cs:
```diff
@@ -1,14 +1,19 @@
-namespace Chapter1_1_2 {
-
-    //構造体
+    
```
ここにも余分なスペースがあります。（ここはすでに2行目に空行があるので、改行自体不要かと思います。）
定期的にCtrl + K > Ctrl + D でコード整形をする癖をつけるとよいです。

teshima-being commented on Chapter1/Chapter1-1-2/Program.cs:
```diff
@@ -1,72 +1,78 @@
-using System.Linq;
-using System.Text;
-using System.Threading.Tasks;
-
 namespace Chapter1_1_2 {
```
usingディレクティブとnamespaceの間には1行改行がある方が見やすいです。
新規でプロジェクトを作成したときも、ここにはあらかじめ1行空行が入っているかと思います。

teshima-being commented on Chapter1/Chapter1-1-3/Program.cs:
```diff
@@ -1,45 +1,34 @@
-            Console.WriteLine("{0}は{1}年{2}組に所属しており、生年月日は{3}です。{4}歳です",
-                wStudent1.Name, wStudent1.Grade, wStudent1.ClassNumber, wStudent1.Birthday, wAge);
-
+            Console.WriteLine($"{wStudent1.Name}は{wStudent1.Grade}年{wStudent1.ClassNumber}組に所属しており、生年月日は{wStudent1.GetBirthdayFormat()}です。{wAge}歳です");
```
※参考（修正は任意）
文字列補間式を使うことで、読みやすくなりましたね！
今回は生年月日の表示形式を指定するためにメソッドを作成してくれていますが、下記のように記述すれば指定した書式で出力することができます。また試してみてください！
`…生年月日は{wStudent1.Birthday:yyyy年MM月dd日}です。…`

teshima-being reviewed (commented):
とても見やすくなりました！
あと少しだけ修正をお願いします。

sanada-being committed:
#516375　[C#テキスト] 1章演習問題

・プロパティをクラス内で参照する場合はthisを付けるように修正しました。
・Chapter1/Chapter1-1-1/Program.cs6行目の余計な改行を削除しました。
・Chapter1/Chapter1-1-1/Program.cs18行目にて、varを用いた記述に変更しました。
・Chapter1/Chapter1-1-1/Product.csにてコンストラクタのオブジェクトコメントを分かりやすい記述に変更しました。
・Chapter1/Chapter1-1-2/Program.csにて、usingディレクティブとnamespaceの間に1行改行を挿入しました。
・Chapter1/Chapter1-1-2/MyStruct.cs3行目の余計なスペースを削除しました。

teshima-being commented on Chapter1/Chapter1-1-3/Person.cs:
```diff
@@ -0,0 +1,36 @@
+﻿using System;
+namespace Chapter1_1_3 {
```
ここも1行空けておきましょう。
`Chapter1-1-3`の`Program.cs`と`Student.cs`も同様です。

teshima-being commented on Chapter1/Chapter1-1-3/Person.cs:
```diff
@@ -0,0 +1,36 @@
+        /// <summary>
+        /// 年齢を取得する。
+        /// </summary>
+        /// <returns>年齢を整数値で返す。</returns>
```
ここもコードからすぐ分かる情報（整数値で）は削除しておきましょうか。
`Chapter1-1-1`の`Product.cs`にある`GetTax()`メソッドや`GetPriceIncludingTax()`メソッドのドキュメントコメントも修正するところがないか確認してみてください。

teshima-being reviewed (commented):
あとちょっとだけ修正をお願いします。
ある箇所で指摘されたことについては他にも同様に修正できるところがないか全体を再確認してみてください。

sanada-being committed:
#517162 [C#テキスト]3章演習問題

・問題回答用のプロジェクトを作成しました。

sanada-being committed:
Revert "#517162 [C#テキスト]3章演習問題"

This reverts commit 17c48d0354d07ff7553da88a2aa35f36ee4e04ae.

sanada-being committed:
#516375 [C#テキスト]1章演習問題

・usingとnamespaceの間に1行改行を挿入しました。
・ドキュメントコメントから、「整数値で」のような見たらわかる表記を削除しました。
・Chapter1/Chapter1-1-3/Person.cs7行目にて、//の後に半角スペースを挿入しました。
・struct.csとClass.csの改行の差を修正しました。

teshima-being commented:
レビューOKです。1章お疲れ様でした。
